File: src/lib/otpauth-uri.test.ts
import { describe, expect, it } from 'vitest';
import { isOtpauthUri, OtpauthUriError, parseOtpauthUri } from './otpauth-uri';

describe('isOtpauthUri', () => {
  it('erkennt otpauth-URIs', () => {
    expect(isOtpauthUri('otpauth://totp/Test?secret=ABCD')).toBe(true);
    expect(isOtpauthUri('  OTPAUTH://TOTP/Test?secret=ABCD  ')).toBe(true);
  });

  it('erkennt alles andere nicht als URI', () => {
    expect(isOtpauthUri('JBSWY3DPEHPK3PXP')).toBe(false);
    expect(isOtpauthUri('https://example.com')).toBe(false);
    expect(isOtpauthUri('otpauth:totp/Test')).toBe(false); // ohne "//"
    expect(isOtpauthUri('')).toBe(false);
  });
});

describe('parseOtpauthUri — URIs, wie echte Anbieter sie erzeugen', () => {
  it('GitHub', () => {
    const parsed = parseOtpauthUri(
      'otpauth://totp/GitHub:kevin?secret=JBSWY3DPEHPK3PXP&issuer=GitHub',
    );
    expect(parsed).toEqual({
      secret: 'JBSWY3DPEHPK3PXP',
      issuer: 'GitHub',
      accountName: 'kevin',
      algorithm: 'SHA-1',
      digits: 6,
      period: 30,
    });
  });

  it('Google (Label vollständig prozent-codiert)', () => {
    const parsed = parseOtpauthUri(
      'otpauth://totp/Google%3Akevin%40gmail.com?secret=JBSWY3DPEHPK3PXP&issuer=Google',
    );
    expect(parsed.issuer).toBe('Google');
    expect(parsed.accountName).toBe('kevin@gmail.com');
    expect(parsed.secret).toBe('JBSWY3DPEHPK3PXP');
  });

  it('Microsoft (alle Parameter explizit)', () => {
    const parsed = parseOtpauthUri(
      'otpauth://totp/Microsoft:kevin%40outlook.com' +
        '?secret=JBSWY3DPEHPK3PXP&issuer=Microsoft&algorithm=SHA1&digits=6&period=30',
    );
    expect(parsed).toEqual({
      secret: 'JBSWY3DPEHPK3PXP',
      issuer: 'Microsoft',
      accountName: 'kevin@outlook.com',
      algorithm: 'SHA-1',
      digits: 6,
      period: 30,
    });
  });

  it('AWS (Kontoname enthält ein @ mit Zahlenkonto)', () => {
    const parsed = parseOtpauthUri(
      'otpauth://totp/AWS:kevin@123456789012?secret=JBSWY3DPEHPK3PXP&issuer=AWS',
    );
    expect(parsed.issuer).toBe('AWS');
    expect(parsed.accountName).toBe('kevin@123456789012');
  });

  it('Issuer mit Leerzeichen', () => {
    const parsed = parseOtpauthUri(
      'otpauth://totp/ACME%20Co:john.doe%40email.com' +
        '?secret=HXDMVJECJJWSRB3HWIZR4IFUGFTMXBOZ&issuer=ACME%20Co&algorithm=SHA1&digits=6&period=30',
    );
    expect(parsed.issuer).toBe('ACME Co');
    expect(parsed.accountName).toBe('john.doe@email.com');
    expect(parsed.secret).toBe('HXDMVJECJJWSRB3HWIZR4IFUGFTMXBOZ');
  });

  it('Label ganz ohne Issuer', () => {
    const parsed = parseOtpauthUri('otpauth://totp/kevin%40example.com?secret=JBSWY3DPEHPK3PXP');
    expect(parsed.issuer).toBeUndefined();
    expect(parsed.accountName).toBe('kevin@example.com');
  });

  it('leeres Label', () => {
    const parsed = parseOtpauthUri('otpauth://totp/?secret=JBSWY3DPEHPK3PXP');
    expect(parsed.issuer).toBeUndefined();
    expect(parsed.accountName).toBeUndefined();
  });

  it('Leerzeichen nach dem Doppelpunkt im Label', () => {
    const parsed = parseOtpauthUri('otpauth://totp/GitHub:%20kevin?secret=JBSWY3DPEHPK3PXP');
    expect(parsed.accountName).toBe('kevin');
  });
});

describe('parseOtpauthUri — Parameter', () => {
  const base = 'otpauth://totp/Test?secret=JBSWY3DPEHPK3PXP';

  it('setzt die Voreinstellungen SHA-1 / 6 / 30, wenn nichts angegeben ist', () => {
    const parsed = parseOtpauthUri(base);
    expect(parsed.algorithm).toBe('SHA-1');
    expect(parsed.digits).toBe(6);
    expect(parsed.period).toBe(30);
  });

  it('übersetzt die Algorithmus-Namen in die Schreibweise der Web Crypto API', () => {
    expect(parseOtpauthUri(`${base}&algorithm=SHA1`).algorithm).toBe('SHA-1');
    expect(parseOtpauthUri(`${base}&algorithm=sha1`).algorithm).toBe('SHA-1');
    expect(parseOtpauthUri(`${base}&algorithm=SHA-1`).algorithm).toBe('SHA-1');
    expect(parseOtpauthUri(`${base}&algorithm=SHA256`).algorithm).toBe('SHA-256');
    expect(parseOtpauthUri(`${base}&algorithm=sha512`).algorithm).toBe('SHA-512');
  });

  it('liest abweichende Stellenzahlen und Perioden', () => {
    expect(parseOtpauthUri(`${base}&digits=8`).digits).toBe(8);
    expect(parseOtpauthUri(`${base}&period=60`).period).toBe(60);
    expect(parseOtpauthUri(`${base}&period=15`).period).toBe(15);
  });

  it('bevorzugt den issuer-Parameter gegenüber dem Label-Präfix', () => {
    // Die Key-Uri-Spezifikation erklärt den Parameter für verbindlich.
    const parsed = parseOtpauthUri(
      'otpauth://totp/Alt:kevin?secret=JBSWY3DPEHPK3PXP&issuer=Richtig',
    );
    expect(parsed.issuer).toBe('Richtig');
    expect(parsed.accountName).toBe('kevin');
  });
});

describe('parseOtpauthUri — Fehlerfälle', () => {
  it('lehnt fremde Schemata ab', () => {
    expect(() => parseOtpauthUri('https://example.com/?secret=ABCD')).toThrow(OtpauthUriError);
  });

  it('lehnt HOTP mit einer erklärenden Meldung ab', () => {
    expect(() => parseOtpauthUri('otpauth://hotp/Test?secret=ABCD&counter=1')).toThrow(/HOTP/);
  });

  it('lehnt unbekannte Typen ab', () => {
    expect(() => parseOtpauthUri('otpauth://xotp/Test?secret=ABCD')).toThrow(/totp/i);
  });

  it('verlangt den Parameter secret', () => {
    expect(() => parseOtpauthUri('otpauth://totp/Test')).toThrow(/secret/);
    expect(() => parseOtpauthUri('otpauth://totp/Test?secret=')).toThrow(/secret/);
    expect(() => parseOtpauthUri('otpauth://totp/Test?issuer=Foo')).toThrow(/secret/);
  });

  it('lehnt unbekannte Algorithmen ab', () => {
    expect(() => parseOtpauthUri('otpauth://totp/T?secret=ABCD&algorithm=MD5')).toThrow(/MD5/);
  });

  it('lehnt unsinnige Zahlenwerte ab', () => {
    expect(() => parseOtpauthUri('otpauth://totp/T?secret=ABCD&digits=5')).toThrow(/digits/);
    expect(() => parseOtpauthUri('otpauth://totp/T?secret=ABCD&digits=9')).toThrow(/digits/);
    expect(() => parseOtpauthUri('otpauth://totp/T?secret=ABCD&digits=sechs')).toThrow(/digits/);
    expect(() => parseOtpauthUri('otpauth://totp/T?secret=ABCD&digits=6x')).toThrow(/digits/);
    expect(() => parseOtpauthUri('otpauth://totp/T?secret=ABCD&period=0')).toThrow(/period/);
    expect(() => parseOtpauthUri('otpauth://totp/T?secret=ABCD&period=-30')).toThrow(/period/);
  });

  it('lehnt kaputte Prozent-Codierung im Label ab', () => {
    expect(() => parseOtpauthUri('otpauth://totp/Test%ZZ?secret=ABCD')).toThrow(/Codierung/);
  });

  it('lehnt komplett unlesbare Eingaben ab', () => {
    expect(() => parseOtpauthUri('das ist keine uri')).toThrow(OtpauthUriError);
  });
});

File: android-native/app/src/test/kotlin/io/github/keco216/clockwork/core/OtpauthUriTest.kt
package io.github.keco216.clockwork.core

import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * Portiert aus `src/lib/otpauth-uri.test.ts`.
 */
class OtpauthUriTest {

    @Test
    fun `isOtpauthUri erkennt otpauth-URIs`() {
        assertTrue(isOtpauthUri("otpauth://totp/Test?secret=ABCD"))
        assertTrue(isOtpauthUri("  OTPAUTH://TOTP/Test?secret=ABCD  "))
    }

    @Test
    fun `isOtpauthUri erkennt alles andere nicht als URI`() {
        assertFalse(isOtpauthUri("JBSWY3DPEHPK3PXP"))
        assertFalse(isOtpauthUri("https://example.com"))
        assertFalse(isOtpauthUri("otpauth:totp/Test")) // ohne "//"
        assertFalse(isOtpauthUri(""))
    }

    @Test
    fun `GitHub`() {
        val parsed = parseOtpauthUri("otpauth://totp/GitHub:kevin?secret=JBSWY3DPEHPK3PXP&issuer=GitHub")
        assertEquals(
            ParsedOtpauthUri("JBSWY3DPEHPK3PXP", "GitHub", "kevin", HashAlgorithm.SHA1, 6, 30),
            parsed,
        )
    }

    @Test
    fun `Google mit vollstaendig prozent-codiertem Label`() {
        val parsed = parseOtpauthUri(
            "otpauth://totp/Google%3Akevin%40gmail.com?secret=JBSWY3DPEHPK3PXP&issuer=Google",
        )
        assertEquals("Google", parsed.issuer)
        assertEquals("kevin@gmail.com", parsed.accountName)
        assertEquals("JBSWY3DPEHPK3PXP", parsed.secret)
    }

    @Test
    fun `Microsoft mit allen Parametern explizit`() {
        val parsed = parseOtpauthUri(
            "otpauth://totp/Microsoft:kevin%40outlook.com" +
                "?secret=JBSWY3DPEHPK3PXP&issuer=Microsoft&algorithm=SHA1&digits=6&period=30",
        )
        assertEquals(
            ParsedOtpauthUri(
                "JBSWY3DPEHPK3PXP", "Microsoft", "kevin@outlook.com", HashAlgorithm.SHA1, 6, 30,
            ),
            parsed,
        )
    }

    @Test
    fun `AWS mit At-Zeichen im Kontonamen`() {
        val parsed = parseOtpauthUri(
            "otpauth://totp/AWS:kevin@123456789012?secret=JBSWY3DPEHPK3PXP&issuer=AWS",
        )
        assertEquals("AWS", parsed.issuer)
        assertEquals("kevin@123456789012", parsed.accountName)
    }

    @Test
    fun `Issuer mit Leerzeichen`() {
        val parsed = parseOtpauthUri(
            "otpauth://totp/ACME%20Co:john.doe%40email.com" +
                "?secret=HXDMVJECJJWSRB3HWIZR4IFUGFTMXBOZ&issuer=ACME%20Co" +
                "&algorithm=SHA1&digits=6&period=30",
        )
        assertEquals("ACME Co", parsed.issuer)
        assertEquals("john.doe@email.com", parsed.accountName)
        assertEquals("HXDMVJECJJWSRB3HWIZR4IFUGFTMXBOZ", parsed.secret)
    }

    @Test
    fun `Label ganz ohne Issuer`() {
        val parsed = parseOtpauthUri("otpauth://totp/kevin%40example.com?secret=JBSWY3DPEHPK3PXP")
        assertNull(parsed.issuer)
        assertEquals("kevin@example.com", parsed.accountName)
    }

    @Test
    fun `leeres Label`() {
        val parsed = parseOtpauthUri("otpauth://totp/?secret=JBSWY3DPEHPK3PXP")
        assertNull(parsed.issuer)
        assertNull(parsed.accountName)
    }

    @Test
    fun `Leerzeichen nach dem Doppelpunkt im Label`() {
        val parsed = parseOtpauthUri("otpauth://totp/GitHub:%20kevin?secret=JBSWY3DPEHPK3PXP")
        assertEquals("kevin", parsed.accountName)
    }

    private val base = "otpauth://totp/Test?secret=JBSWY3DPEHPK3PXP"

    @Test
    fun `setzt die Voreinstellungen SHA-1 6 30`() {
        val parsed = parseOtpauthUri(base)
        assertEquals(HashAlgorithm.SHA1, parsed.algorithm)
        assertEquals(6, parsed.digits)
        assertEquals(30, parsed.period)
    }

    @Test
    fun `uebersetzt die Algorithmus-Namen`() {
        assertEquals(HashAlgorithm.SHA1, parseOtpauthUri("$base&algorithm=SHA1").algorithm)
        assertEquals(HashAlgorithm.SHA1, parseOtpauthUri("$base&algorithm=sha1").algorithm)
        assertEquals(HashAlgorithm.SHA1, parseOtpauthUri("$base&algorithm=SHA-1").algorithm)
        assertEquals(HashAlgorithm.SHA256, parseOtpauthUri("$base&algorithm=SHA256").algorithm)
        assertEquals(HashAlgorithm.SHA512, parseOtpauthUri("$base&algorithm=sha512").algorithm)
    }

    @Test
    fun `liest abweichende Stellenzahlen und Perioden`() {
        assertEquals(8, parseOtpauthUri("$base&digits=8").digits)
        assertEquals(60, parseOtpauthUri("$base&period=60").period)
        assertEquals(15, parseOtpauthUri("$base&period=15").period)
    }

    @Test
    fun `bevorzugt den issuer-Parameter gegenueber dem Label-Praefix`() {
        // Die Key-Uri-Spezifikation erklaert den Parameter fuer verbindlich.
        val parsed = parseOtpauthUri("otpauth://totp/Alt:kevin?secret=JBSWY3DPEHPK3PXP&issuer=Richtig")
        assertEquals("Richtig", parsed.issuer)
        assertEquals("kevin", parsed.accountName)
    }

    @Test
    fun `lehnt fremde Schemata ab`() {
        val error = capture { parseOtpauthUri("https://example.com/?secret=ABCD") }
        assertEquals("err.uri.scheme", error.key)
        assertEquals("https", error.args["scheme"])
    }

    @Test
    fun `lehnt HOTP mit einer erklaerenden Meldung ab`() {
        assertKey("err.uri.hotp") { parseOtpauthUri("otpauth://hotp/Test?secret=ABCD&counter=1") }
    }

    @Test
    fun `lehnt unbekannte Typen ab`() {
        val error = capture { parseOtpauthUri("otpauth://xotp/Test?secret=ABCD") }
        assertEquals("err.uri.type", error.key)
        assertEquals("xotp", error.args["type"])
    }

    @Test
    fun `verlangt den Parameter secret`() {
        assertKey("err.uri.noSecret") { parseOtpauthUri("otpauth://totp/Test") }
        assertKey("err.uri.noSecret") { parseOtpauthUri("otpauth://totp/Test?secret=") }
        assertKey("err.uri.noSecret") { parseOtpauthUri("otpauth://totp/Test?issuer=Foo") }
    }

    @Test
    fun `lehnt unbekannte Algorithmen ab`() {
        val error = capture { parseOtpauthUri("otpauth://totp/T?secret=ABCD&algorithm=MD5") }
        assertEquals("err.uri.algorithm", error.key)
        assertEquals("MD5", error.args["value"])
    }

    @Test
    fun `lehnt unsinnige Zahlenwerte ab`() {
        assertKey("err.uri.digits") { parseOtpauthUri("otpauth://totp/T?secret=ABCD&digits=5") }
        assertKey("err.uri.digits") { parseOtpauthUri("otpauth://totp/T?secret=ABCD&digits=9") }
        assertKey("err.uri.integer") { parseOtpauthUri("otpauth://totp/T?secret=ABCD&digits=sechs") }
        assertKey("err.uri.integer") { parseOtpauthUri("otpauth://totp/T?secret=ABCD&digits=6x") }
        assertKey("err.uri.period") { parseOtpauthUri("otpauth://totp/T?secret=ABCD&period=0") }
        // "-30" faellt schon am Ziffernmuster durch, nicht erst am Bereich —
        // genau wie im Web, wo /^\d+$/ vor der Bereichspruefung greift.
        assertKey("err.uri.integer") { parseOtpauthUri("otpauth://totp/T?secret=ABCD&period=-30") }
    }

    @Test
    fun `lehnt kaputte Prozent-Codierung im Label ab`() {
        assertKey("err.uri.badLabel") { parseOtpauthUri("otpauth://totp/Test%ZZ?secret=ABCD") }
    }

    @Test
    fun `lehnt komplett unlesbare Eingaben ab`() {
        assertKey("err.uri.invalid") { parseOtpauthUri("das ist keine uri") }
    }

    @Test
    fun `ein Plus im Parameter bleibt ein Plus`() {
        // Bewusster Unterschied zur Web-Fassung, siehe PercentCodec: Dort
        // liest `URLSearchParams` das `+` als Leerzeichen, ein Anbieter "AT+T"
        // hiesse also "AT T". Hier bleibt er stehen.
        val parsed = parseOtpauthUri("$base&issuer=AT%2BT")
        assertEquals("AT+T", parsed.issuer)
    }
}
